statsmodels/genmod/cov_struct.py

```diff
@@ -951,36 +972,130 @@ class Equivalence(CovStruct):
     >> pairs[1][2] = 3 + np.tril_indices(3, -1)
     """
 
-    def __init__(self, pairs):
+    def __init__(self, pairs=None, labels=None, as_cor=True):
 
         super(Equivalence, self).__init__()
 
+        if (pairs is None) and (labels is None):
+            raise ValueError("Equivalence cov_struct requires either `pairs` or `labels`")
+
+        if (pairs is not None) and (labels is not None):
+            raise ValueError("Equivalence cov_struct accepts only one of `pairs` and `labels`")
+
+        if pairs is not None:
+            self.pairs = pairs
+        if labels is not None:
+            self.labels = np.asarray(labels)
+
+        self.as_cor = as_cor
+
+    def _make_pairs(self, i, j):
+        """
+        Create arrays `i_`, `j_` containing all unique ordered pairs of elements in `i` and `j`.
+
+        The arrays `i` and `j` must be one-dimensional containing non-negative integers.
+        """
+
+        mat = np.zeros((len(i)*len(j), 2), dtype=np.int32)
+
+        # Create the pairs and order them
+        f = np.ones(len(j))
+        mat[:, 0] = np.kron(f, i).astype(np.int32)
+        f = np.ones(len(i))
+        mat[:, 1] = np.kron(j, f).astype(np.int32)
+        mat.sort(1)
+
+        # Remove repeated rows
+        try:
+            dtype = np.dtype((np.void, mat.dtype.itemsize * mat.shape[1]))
+            bmat = np.ascontiguousarray(mat).view(dtype)
+            _, idx = np.unique(bmat, return_index=True)
+        except TypeError:
+            # workaround for old numpy that can't call unique with complex
+            # dtypes
```

Comment:
I think it's only an issue because I am applying unique to what is
effectively a very long record constructed by concatenating all the bytes
in one row of an ndarray.  I'm not sure about the details, but there is
some discussion here:

http://stackoverflow.com/questions/16970982/find-unique-rows-in-numpy-array

On Wed, Jan 21, 2015 at 9:55 PM, Josef Perktold <notifications@github.com>
wrote:

> In statsmodels/genmod/cov_struct.py
> <https://github.com/statsmodels/statsmodels/pull/2188#discussion_r23352751>
> :
>
> > +
> > +        # Create the pairs and order them
> > +        f = np.ones(len(j))
> > +        mat[:, 0] = np.kron(f, i).astype(np.int32)
> > +        f = np.ones(len(i))
> > +        mat[:, 1] = np.kron(j, f).astype(np.int32)
> > +        mat.sort(1)
> > +
> > +        # Remove repeated rows
> > +        try:
> > +            dtype = np.dtype((np.void, mat.dtype.itemsize * mat.shape[1]))
> > +            bmat = np.ascontiguousarray(mat).view(dtype)
> > +            _, idx = np.unique(bmat, return_index=True)
> > +        except TypeError:
> > +            # workaround for old numpy that can't call unique with complex
> > +            # dtypes
>
> why us this part necessary?
> (need to check: there was a problem for 1 or 2 numpy versions with
> return_xxx in np.unique. statsmodels.compat ?)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/statsmodels/statsmodels/pull/2188/files#r23352751>.
>